### DISCUSSION TITLE
Snapshot template that honors SemVer

### DISCUSSION THREAD

**Discussion by zam6ak** [goreleaser/goreleaser#1455] [Ideas, Issue Triage, and General discussions]
I have a simple project with master branch and each release is tagged.
However, I am trying to figure out how name snapshots properly so that it follows SemVer.

Imagine you release (and tag) v1.0.0
The next patch release could undergo several snapshot builds before final v1.0.1 release 
So there will be for example:
v1.0.1-alpha.0 
v1.0.1-alpha.1 
v1.0.1-alpha.2 
and then finally
v1.0.1
which is then tagged in the repo

My issue is that because of snapshot template, I am getting the v1.0.0-alpha.0 and so on because that is the only released tag in the repo

**Is there a way to adjust the template so that it uses proper semver *without* having to tag each snapshot build in the repo?**

This is what I use now:
```
snapshot:
  name_template: "{{ .Tag }}-alpha-{{.ShortCommit}}"
```

I could remove tag from template and mess with Major and/or Minor, but I suspect someone has an easier way...

Thanks

**Comment by zam6ak** [comment]
Would be awsome if one could "apply" a function to .Tag...
Something like (using NPM "semver" command syntax here)

```
snapshot:
  name_template: "{{semver --preid alpha --increment prerelease .Tag }}-{{.ShortCommit}}"
```
This would result (if `v1.0.0` is last tag in repo) in
`1.0.1-alpha.0-4c7af85`

**Reply by caarlos0** [reply]
would this help? https://github.com/goreleaser/goreleaser/pull/1457

**Reply by caarlos0** [reply]
cc/ @zam6ak

**Reply by zam6ak** [reply]
that would help!

**Comment by caarlos0** [comment]
that's not possible right now, but I think we can add template functions like `incmajor`, `incminor` and `incpatch`

**Reply by zam6ak** [reply]
not sure how flexible this would end up being (ir. being able to specify suffix)

**Reply by caarlos0** [reply]
on your example:

```yaml
snapshot:
  name_template: "{{semver --preid alpha --increment prerelease .Tag }}-{{.ShortCommit}}"
```

would be something like

```yaml
snapshot:
  name_template: "{{ .Tag | incpatch }}-alpha.{{.ShortCommit}}"
```

(or something like that)

---

does that work?

**Reply by zam6ak** [reply]
I think it would except in cases of prerelease...How would you specify "moniker" used for "rc" or whatever else people may be using....

**Reply by caarlos0** [reply]
in that case you don't need snapshots... just tag with the 1.2.3-rc.blah and configure `release.prerelease` to `auto`: https://goreleaser.com/release/

**Comment by zam6ak** [comment]
Was just thinking....would it be possible to have values of the template "piped" through external script or command? That way anyone could modify `.Tag` (or any other part of the template) without having to implement new fn each time need arises.

**Reply by caarlos0** [reply]
sorry, I didn't understand what you mean... can you elaborate?

**Reply by zam6ak** [reply]
Let's say you have [semver](https://github.com/npm/node-semver)  utility installed and in your path
Then having function that would allow value of the tag to be used by such external utility
```
{{ .Tag | extexec  "semver --preid alpha --increment prerelease" }}
```

Here you would only implement `extexec` which can be used with whatever external utility someone wants to use to manipulate value of the tag...

**Reply by caarlos0** [reply]
you can do that with

```
SEMVER="$(semver --preid alpha --increment prerelease)" goreleaser
```

and then use `{{ .Env.SEMVER }}` on templates...
